### DISCUSSION TITLE
Force instant scrolling in scroll restoration

### DISCUSSION THREAD

**Discussion by david-bezero** [remix-run/react-router#10328] [Proposals]
Currently if a page uses `ScrollRestoration` and has the following CSS:

```css
html {
  scroll-behavior: smooth
}
```

then the page will appear to jump around in a non-intuitive way when switching to a new page (specifically: it will show the current scroll position on the new page, then animate to the correct scroll position). Having this global smooth scrolling behaviour can be desirable in some cases (e.g. within large pages with internal anchor links), but is never desirable when switching between pages.

I propose changing `ScrollRestoration` to force immediate scrolling (see #10318 for details)

On that PR, it was suggested that an alternative approach might be to make the `scrollTo` function into customisable parameter. This would also enable solving the issue, but since the default behaviour would continue to be surprising in this scenario, it puts the burden on every user individually to fix it.

**Comment by herrwitzi** [comment]
I too find the alternative approach with a custom `scrollTo` function better, because it is not breaking. The burden is on the user to implement a fix by themselves, but I think with good documentation that explicitly mentions this problem with smooth-scrolling, the user will easily find it.

I already submitted a PR [#14134](https://github.com/remix-run/react-router/pull/14134) but got redirected to further advance this discussion.

**Again, my proposal for better visibility:**

Adds an optional `scrollTo` prop to `ScrollRestoration` that is used instead of the default `window.scrollTo` calls. This allows the user to specify a custom function for scroll restoration.

For example overwriting a global CSS scroll-behavior (e.g., smooth on the html tag).

```jsx
<ScrollRestoration
  scrollTo={(y) => {
    document.documentElement.style.scrollBehavior = "auto"
    window.scrollTo(0, y)
    document.documentElement.style.scrollBehavior = ""
  }}
/>
```
